File: Alcoholtracker/Services/ServingSizeMemory.swift
import Foundation

// MARK: - ServingSizeMemory
//
// Remembers the serving size (volume in ml) the user last chose for a given drink
// template, so the next time they open the amount sheet that size ("Dose", "Flasche
// 0,5 L", ...) is pre-selected automatically instead of always defaulting to the
// template's nominal volume. Keyed by templateID, stored in UserDefaults (no
// SwiftData migration needed). A preset is matched back by its volume, so saving the
// chosen volume is enough to restore the chosen format.
enum ServingSizeMemory {

    private static let storageKey = "servingSizeMemory_v1"

    private static var storage: [String: Double] {
        get { UserDefaults.standard.dictionary(forKey: storageKey) as? [String: Double] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: storageKey) }
    }

    /// The user's last chosen volume (ml) for this template, or nil if never set.
    static func volume(for templateID: UUID) -> Double? {
        storage[templateID.uuidString]
    }

    /// Persists the chosen volume so it becomes the default next time.
    static func save(volume: Double, for templateID: UUID) {
        guard volume > 0 else { return }
        var s = storage
        s[templateID.uuidString] = volume
        storage = s
    }
}

File: android/app/src/main/kotlin/de/tipau/promille/service/ServingSizeMemory.kt
package de.tipau.promille.service

import android.content.Context

/**
 * Port of iOS ServingSizeMemory. Remembers the serving size (volume in ml) the
 * user last chose for a given drink template, so the amount sheet pre-selects
 * that size next time instead of always defaulting to the template's nominal
 * volume. Keyed by templateID, stored in SharedPreferences. A preset is
 * matched back by its volume, so saving the chosen volume is enough to
 * restore the chosen format.
 */
object ServingSizeMemory {
    private const val PREFS = "serving_size_memory_v1"

    /** The user's last chosen volume (ml) for this template, or null if never set. */
    fun volume(context: Context, templateID: String): Double? {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (!prefs.contains(templateID)) return null
        val v = prefs.getFloat(templateID, 0f).toDouble()
        return if (v > 0) v else null
    }

    /** Persists the chosen volume so it becomes the default next time. */
    fun save(context: Context, templateID: String, volume: Double) {
        if (volume <= 0) return
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putFloat(templateID, volume.toFloat())
            .apply()
    }
}
